token-erc20/chaincode-javascript/lib/tokenERC20.js

```diff
@@ -0,0 +1,367 @@
+
+        return true;
+    }
+
+    /**
+     * Allows `spender` to spend `value` amount of tokens from the owner.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} spender The spender
+     * @param {Integer} value The amount of tokens to be approved for transfer
+     * @returns {Boolean} Return whether the approval was successful or not
+     */
+    async approve(ctx, spender, value) {
+        const owner = ctx.clientIdentity.getID();
+
+        const allowanceKey = ctx.stub.createCompositeKey(allowancePrefix, [owner, spender]);
+        if (!allowanceKey) {
+            throw new Error('Failed to create the composite key');
+        }
+
+        let valueInt = parseInt(value);
+        await ctx.stub.putState(allowanceKey, Buffer.from(valueInt.toString()));
+
+        // Emit the Approval event
+        const approvalEvent = { owner, spender, value: valueInt };
+        ctx.stub.setEvent('Approval', Buffer.from(JSON.stringify(approvalEvent)));
+
+        console.log('approve ended successfully');
+        return true;
+    }
+
+    /**
+     * Returns the amount of tokens which `spender` is allowed to withdraw from `owner`.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} owner The owner of tokens
+     * @param {String} spender The spender who are able to transfer the tokens
+     * @returns {Number} Return the amount of remaining tokens allowed to spent
+     */
+    async allowance(ctx, owner, spender) {
+        const allowanceKey = ctx.stub.createCompositeKey(allowancePrefix, [owner, spender]);
+        if (!allowanceKey) {
+            throw new Error('Failed to create the composite key');
+        }
+
+        const allowanceBytes = await ctx.stub.getState(allowanceKey);
+        const allowance = parseInt(allowanceBytes.toString());
+        return allowance;
+    }
+
+    // ================== Extended Functions ==========================
+
+    /**
+     * Set optional infomation for a token.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} name The name of the token
+     * @param {String} symbol The symbol of the token
+     * @param {String} decimals The decimals of the token
+     * @param {String} totalSupply The totalSupply of the token
+     */
+    async setOption(ctx, name, symbol, decimals) {
+        await ctx.stub.putState(nameKey, Buffer.from(name));
+        await ctx.stub.putState(symbolKey, Buffer.from(symbol));
+        await ctx.stub.putState(decimalsKey, Buffer.from(decimals));
+
+        console.log(`name: ${name}, symbol: ${symbol}, decimals: ${decimals}`);
+        return true;
+    }
+
+    /**
+     * Mint fungible token
+     *
+     * @param {Context} ctx the transaction context
+     * @param {Integer} amount amount of tokens to be minted
+     * @returns {Object} The balance
+     */
+    async mint(ctx, amount) {
```

Comment:
I agree that only a central banker can mint tokens. I added the same auth check.
As you know, ERC-20 spec does not have mint() function. OpenZeppelin's ERC20 implementation has role-based access control (https://docs.openzeppelin.com/contracts/3.x/access-control). It is beyond the scope of this sample to have the same access control methods because we just want to show how to use tokens. I think it's enough to check MSPID.
I also added the same auth check to Burn() because the tokens should be redeemed by only a central banker.